### PR TITLE
XFA - Add a storage to save fields values

### PR DESCRIPTION
**Pull request by calixteman**
- this is required to be able to print (or save) a document. Some pages can be unloaded (because pdf.js is lazy) and this storage will help to save their data in order to resuse them when printing or just when displaying a page again.

### CODE REVIEW

File: src/display/xfa_layer.js [hunk 1]
```diff
@@ -30,13 +81,20 @@ class XfaLayer {
         Object.assign(html.style, value);
       }
     }
+
+    // Set the value after the others to be sure overwrite
+    // any other values.
+    if ("fieldId" in attributes && storage) {
```
Comment (Snuffleupagus) [inline src/display/xfa_layer.js:87]: Nit: It's probably a tiny bit cheaper to re-order this as `if (storage && attributes.fieldId !== undefined) {` instead.

File: src/display/xfa_layer.js [hunk 2]
```diff
@@ -30,13 +81,20 @@ class XfaLayer {
...
+      XfaLayer.setupStorage(html, attributes.fieldId, element, storage);
```
Comment (Snuffleupagus) [inline src/display/xfa_layer.js:88]: Nit: Here, and elsewhere in this file, note that `static` methods should be able to call each other directly; hence `this.setupStorage(...)` ought to work.

**Comment by brendandahl** [comment]
Will this be needed if we keep all pages in cache?

**Comment by Snuffleupagus** [comment]
> Will this be needed if we keep all pages in cache?

That would be breaking some very old/basic assumptions in the viewer, since only one-page documents are ever guaranteed to be fully rendered at all times. Relaxing that assumption could be quite bad in the general case, for e.g. memory usage and performance, for anything except very short documents and as far as I can tell there's no *guarantee* that all XFA-documents are indeed "short".

**Review by brendandahl — APPROVED** [review]

**Pull request merged**